### ISSUE TITLE
Neural Architecture Search TL;DR

### ISSUE THREAD

**Issue by howardyclo** [howardyclo/papernotes#55]
## Neural Architecture Search TL;DR
This TL;DR aims to help myself getting a quick view of neural architecture search literature. Hopes it helps you too 😁! Contributions are welcomed.

> Designing neural network for task of interest often requires significant architecture engineering. Neural Architecture Search (NAS/AutoML) aims to automatically find a neural network with an optimal architecture such that it can achieve good or even state-of-the-art performance on the task of interest. Most of the NAS work focus on designing the search space (what activation functions or cell operations to search on). Two main categories of NAS: Evolutionary (ES) algorithms and Reinforcment learning (RL) algorithms. Here we mainly overview RL-based methods.

### Reading List
- [x] NAS
- [x] NASNet
- [x] EAS
- [ ] MorphNet
- [x] PNAS
- [ ] DPP-net
- [X] ENAS
- [ ] NetAdapt
- [ ] Understanding and Simplifying One-Shot Architecture Search
- [ ] DARTS
- [ ] InstaNAS
- [ ] ProxylessNAS
- [ ] AutoDeepLab
- [ ] Single-Path NAS

### [1] [Neural Architecture Search with Reinforcement Learning (NAS)](https://arxiv.org/pdf/1611.01578.pdf). ICLR 2017.
![](https://i.imgur.com/LVvsrja.png)
![](https://i.imgur.com/19eEhEh.png)
- Uses a RNN controller to output variable-length tokens (i.e., search space), which specifies a neural network.
- Such neural network is then built, trained, and yeild validation accuracy on validation set.
- Train RNN with this validation accuracy as reward using policy gradient, to encourage searched network generalized well instead of overfitting on training set. (This is essentially meta-learning)
- Also modified the RNN controller to output recursively to search the operations for hidden states in RNN/LSTM cell.
- Trained on 800 GPUs and searched ~15000 architectures over 28 days.
- Evaluated on CIFAR-10 (3.65% error rate) and PTB dataset (word:62.4 ppl; char:1.214 ppl).

### [2] [Learning Transferable Architectures for Scalable Image Recognition (NASNet)](https://arxiv.org/pdf/1707.07012.pdf). CVPR 2018.
![](https://i.imgur.com/mI4BR1a.png)
- Address large computation when directly search on large dataset (e.g., ImageNet): Modify the search space to search for the best cell structure instead of an entire network structure, and may better generalize to other tasks (called *NASNet search space*).
- Similar to RNN/LSTM cell [1], first predict which two input hidden states and then predicts what operations to apply on them.
- The number of cell stacking is considered as hyperparameter tailored to the scale of network.
- They first search on CIFAR-10 (called *NASNets*) and transfer to ImageNet without much modification.
- The optimization is same as [1] but replace vanilla policy gradient as *proximal policy optimization (ppo)* method.
- Trained on 500 GPUs over 4 days and is 7x faster than [1].
- Evaluated on CIFAR-10 (2.4% error rate) and ImageNet (82.7% top-1 acc = SENet)
- The image feature learned by NASNets are generally useful and transfer to other computer vision tasks (e.g., object detection achieves 43.1% mAP on COCO).

### [3] [Efficient Architecture Search by Network Transformation (EAS)](https://arxiv.org/pdf/1707.04873.pdf). AAAI 2018. [[Code](https://github.com/han-cai/EAS)]
![](https://i.imgur.com/ZLQvWHk.png)
- Instead of search and train an network from scratch repeatly, they propose to search an *network transformation operation* (Net2Net) given the current searched network.
- [Net2Net operations](https://arxiv.org/pdf/1511.05641.pdf): Net2Wider (replace a layer with a wider layer) & Net2Deeper (insert a new layer).
- The bi-LSTM takes current architecture as consideration and output on which Net2Net operation to apply.
- Trained on 5 GPUs less than 2 days.
- Evaluated on CIFAR-10+ (4.23% error rate) and SVHN (1.83% error rate).

### [4] [Progressive Neural Architecture Search (PNAS)](https://arxiv.org/pdf/1712.00559.pdf). ECCV 2018.
![](https://i.imgur.com/tL2xk4W.png)
- Built on [2], they propose a sequential model-based optimization (SMBO) strategy (from training simple network to complex one) which is 5x more efficient and 8x computationally faster than [2]. (However they empirically reduce search space compared to [2])
- Begin by training all 1-block cells. They are only 136 (unique) combinations.
- Pick K most promising cells expand into 2-block cells, and iterate again (up to 5-block). However cell structures of 2-block has ~10^5 combinations.
- So we train a surrogate model (MLP or LSTM) predicts the final performance of a cell structure, to help pick promising 2-block cell structures.
- Evaluated on CIFAR-10 (3.41% test error) and ImageNet (82.9% top-1 acc).

### [5] [Efficient Neural Architecture Search via Parameter Sharing (ENAS)](https://arxiv.org/pdf/1802.03268.pdf). ICML 2018. [[TF Code](https://github.com/chenxi116/PNASNet.TF)][[PT Code](https://github.com/chenxi116/PNASNet.pytorch)]
![](https://i.imgur.com/TzTfrpD.png)
- Similar to [3], search an optimal subgraph in current large computational graph (neural network).
- ENAS allows parameters to be shared by representing NAS's search space using a single directed acyclic gragh (DAG), where an architecture can be realized by taking a subgraph of the DAG (node = local computation; edge = information flow; each pair of node has a weight parameter matrix).
- ENAS controller decides: 1) which edge are activated; 2) which computation are performed at each node. This allows ENAS to design both topology and operation in RNN cell, which is different from [1].
- Alternative training: 1) For a minibatch, train *K* network sampled from the ENAS controller. *K* = 1 works. Train after an entire epoch; 2) Train ENAS controller with policy gradient like [1].
- Deriving final architecture: Sample M network, evaluate on a batch of validation set, select the network that gets best performance and retrain from scratch.
- **Trained on 1 GPU less than 16 hours.**
- Evaluated on CIFAR-10 (2.89% error rate) PTB (word 55.8 ppl).
- Worst a bit than [1] since [1] explores different (more) topologies, while ENAS search on predefined topology.
